apollo-router/src/plugins/telemetry/filtering_layer.rs

```diff
@@ -0,0 +1,141 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+
+use tracing::field;
+use tracing::Level;
+use tracing::Subscriber;
+use tracing_core::Field;
+use tracing_subscriber::Layer;
+
+// Specific attributes for logging
+pub(crate) const SPECIFIC_ATTRIBUTES: [&str; 4] = [
+    "request",
+    "response_headers",
+    "response_body",
+    "operation_name",
+];
+
+const SUBGRAPH_ATTRIBUTE_NAME: &str = "subgraph";
+
+#[derive(Debug, Clone, Default, PartialEq)]
+pub(crate) struct AttributesToExclude {
+    pub(crate) supergraph: HashSet<String>,
+    pub(crate) all_subgraphs: HashSet<String>,
+    pub(crate) subgraphs: HashMap<String, HashSet<String>>,
+}
+
+pub(crate) struct FilterLayer {
+    attributes_to_exclude: Option<AttributesToExclude>,
+}
+
+impl FilterLayer {
+    pub(crate) fn new(attributes_to_exclude: Option<AttributesToExclude>) -> Self {
+        Self {
+            attributes_to_exclude,
+        }
+    }
+}
+
+impl<S> Layer<S> for FilterLayer
+where
+    S: Subscriber + Send + Sync + 'static,
+{
+    fn enabled(
+        &self,
+        metadata: &tracing::Metadata<'_>,
+        ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        // The logic is splitted between `enabled` and `event_enabled` because we can't evaluate the value of an attribute in this method
+        // This method is evaluated only once for a specific callsite, which means if I return false here I won't re-enter in this method everytime for this event
+        // Which is better for performance. So for supergraph level we can use this method, for subgraph level we have to know the value of subgraph attribute
+        if metadata.level() == &Level::DEBUG && metadata.is_event() {
+            match &self.attributes_to_exclude {
+                Some(attributes_to_exclude) => {
+                    let mut subgraph = false;
+                    let mut to_exclude = false;
+                    for field_name in metadata.fields().iter().map(|f| f.name()) {
+                        if field_name == SUBGRAPH_ATTRIBUTE_NAME {
+                            subgraph = true;
+                            // Cannot do anything for subgraph at this level. It will be in `event_enabled` method.
+                            break;
+                        }
+                        if attributes_to_exclude.supergraph.contains(field_name) {
+                            to_exclude = true;
+                        }
+                    }
+                    if !subgraph && to_exclude {
```

Comment:
so there's something that's not clear to me here: is the goal of that feature to add/remove some attributes from logs, or to remove logs containing some specific attributes? If it's the latter, could we rename the  `include_attributes` configuration option to `contains_attributes` or `has_attributes`? Because `include_attributes` feels a bit like "add these attributes to the log messages"

Reply:
Yes it doesn't show the entire event. Because the tracing API doesn't let us to filter or to hide specific attributes on an event. The way to do that is to use a custom formatter, but if we do this, it means for JSON formatter we will have to copy/paste the code of JSON formatter from tracing codebase and it's a big chunk of code. BUT... in the future... I already opened a PR to add new feature on the tracing API to be able to filter attributes. But it's a more long term solution because you know, waiting for the PR to be merged, and then for the release and so one ....